boot-qemu.py

```diff
@@ -0,0 +1,730 @@
+    Prints the QEMU command in a "pretty" manner, similar to how 'set -x' works in bash.
+        * Surrounds list elements that have spaces with quotation marks so that
+          copying and pasting the command in a shell works.
+        * Prints the QEMU executable as just the executable name, rather than
+          the full path. This is done purely for aesthetic reasons, as the
+          executable would normally be called with just its name through PATH
+          but subprocess.Popen() recommends using a full path for maximum
+          compatibility so it was generated in get_qemu_args().
+
+    Parameters:
+        qemu_cmd (list): QEMU command list.
+    """
+    qemu_cmd_pretty = ""
+    for element in qemu_cmd:
+        if " " in element:
+            qemu_cmd_pretty += ' "{}"'.format(element)
+        elif "qemu-system-" in element:
+            qemu_cmd_pretty += " {}".format(element.split("/")[-1])
+        else:
+            qemu_cmd_pretty += " {}".format(element)
+    print("$ {}".format(qemu_cmd_pretty.strip()))
+
+
+def launch_qemu(cfg):
+    """
+    Runs the QEMU command generated from get_qemu_args(), depending on whether
+    or not the user wants to debug with GDB.
+
+    If debugging with GDB, QEMU is called with '-s -S' in the background then
+    gdb_bin is called against 'vmlinux' connected to the target remote. This
+    can be repeated multiple times.
+
+    Otherwise, QEMU is called with 'timeout' so that it is terminated if there
+    is a problem while booting, passing along any error code that is returned.
+
+    Parameters:
+        cfg (dict): The configuration dictionary generated with setup_cfg().
+    """
+    interactive = cfg["interactive"]
+    gdb = cfg["gdb"]
+    gdb_bin = cfg["gdb_bin"]
+    kernel_location = cfg["kernel_location"]
+    qemu_cmd = cfg["qemu_cmd"]
+    timeout = cfg["timeout"]
+
+    # Print information about the QEMU binary
+    pretty_print_qemu_info(qemu_cmd[0])
+
+    if gdb:
+        while True:
+            utils.check_cmd("lsof")
+            lsof = subprocess.run(["lsof", "-i:1234"],
+                                  stdout=subprocess.DEVNULL,
+                                  stderr=subprocess.DEVNULL)
+            if lsof.returncode == 0:
+                utils.die("Port 1234 is already in use, is QEMU running?")
+
+            utils.green("Starting QEMU with GDB connection on port 1234...")
+            qemu_process = subprocess.Popen(qemu_cmd + ["-s", "-S"])
+
+            utils.green("Starting GDB...")
+            utils.check_cmd(gdb_bin)
+            gdb_cmd = [gdb_bin]
+            gdb_cmd += [Path(kernel_location).joinpath("vmlinux").as_posix()]
+            gdb_cmd += ["-ex", "target remote :1234"]
+            subprocess.run(gdb_cmd)
+
+            utils.red("Killing QEMU...")
+            qemu_process.kill()
+            qemu_process.wait()
+
+            answer = input("Re-run QEMU + gdb? [y/n] ")
+            if answer.lower() == "n":
+                exit(0)
```

Comment:
could we simply `break` out of the `while True` here?

Reply:
Sure, I guess that would be better in case we add anything after this if block that is meant to be run unconditionally: 312aaefd30d3aa167c095c9788605d8631d2552e